google/datalab/contrib/bigquery/operators/_bq_extract_operator.py

```diff
@@ -20,23 +20,36 @@ class ExtractOperator(BaseOperator):
   template_fields = ('_table', '_path')
 
   @apply_defaults
-  def __init__(self, table, path, format='csv', csv_options=None, *args, **kwargs):
+  def __init__(self, path, table=None, format='csv', csv_options=None, *args, **kwargs):
     super(ExtractOperator, self).__init__(*args, **kwargs)
     self._table = table
     self._path = path
     self._format = format
     self._csv_options = csv_options or {}
 
   def execute(self, context):
-    if self._table:
-      pydatalab_context = google.datalab.Context.default()
-      source_table = google.datalab.bigquery.Table(self._table, context=pydatalab_context)
-      job = source_table.extract(
-        self._path, format='CSV' if self._format == 'csv' else 'NEWLINE_DELIMITED_JSON',
-        csv_delimiter=self._csv_options.get('delimiter'),
-        csv_header=self._csv_options.get('header'), compress=self._csv_options.get('compress'))
-    else:
-      raise Exception('A table is needed to extract')
+    if not self._table:
+      task_instance = context['task_instance']
+      # If the table is not specified, we fetch it from the output of the execute task, i.e. the
+      # query results table. This could either be a permanent table or a temporary table. If we're
+      # here, it is most likely a temporary table. If it was a permanent one, it would have been
+      # passed in as a param and we wouldn't be here.
+      # TODO(rajivpb): Assert that if we're here, then the table is a temporary one.
+      # TODO(rajivpb):
+      # The task id of the execute task is created by concatenating 'bq_pipeline_execute_task'
+      # and '_id'. This is currently being hard-coded, but consider making this a parameter.
+      # It could require substantial changes to the underlying object model of Pipeline.
+      self._table = task_instance.xcom_pull(task_ids='bq_pipeline_execute_task_id')
+
+    if not self._table:
+      raise Exception('A table is needed to extract.')
```

Comment:
What does it mean to get here?

- A mis-configured pipeline? If so, why not try to catch this at pipeline definition time?
- A failure to have an output table during execute? If so, why does the pipeline execution get to this stage? I would expect execute to fail, and for Airflow to retry that task.

Reply:
This was some defensive/paranoid coding. Removed now.

Reply:
Was looking into this some more and realized that the previous response was incomplete: 

 - If the pipeline is misconfigured, we will indeed catch it at pipeline definition time once the bug on jsonschema validation is fixed: https://github.com/googledatalab/pydatalab/issues/499

 - If there is as failure to have an output table during execute an exception will get thrown downstream, so we should never be here. 
https://github.com/googledatalab/pydatalab/blob/64456d040bb7c856ee99f75bc5a1f179b6dabe0b/google/datalab/bigquery/_query_job.py#L109

Either way, removing the if block and not throwing here makes sense.